src/Paprika/Chain/PooledSpanDictionary.cs

```diff
@@ -0,0 +1,220 @@
+﻿using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using Paprika.Store;
+
+namespace Paprika.Chain;
+
+/// <summary>
+/// The pooled span dictionary uses memory from <see cref="BufferPool"/> to store large chunks of memory.
+/// </summary>
+/// <remarks>
+/// This component hacks into the dictionary to provide the underlying keys by implementing the custom equality comparer.
+/// It allows to store in the underlying dict just simple entries and not pass the reference in there. To get the reference,
+/// the <see cref="Equals"/> gets the underlying <see cref="Span{Byte}"/> from pages;
+/// </remarks>
+public class PooledSpanDictionary : IEqualityComparer<PooledSpanDictionary.KeySpan>, IDisposable
+{
+    private const int BufferSize = BufferPool.BufferSize;
+    private readonly BufferPool _pool;
+    private readonly Dictionary<KeySpan, ValueSpan> _dict;
```

Comment:
Should this dictionary ever be cleared?

Reply:
It's dropped and not reused. If we consider pooling of the dict, then yes. For now, not needed.